crates/dekoder/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+    KeyError(String),
+}
+
+/// My result type has always my errros.
+type Result<T> = std::result::Result<T, EKOError>;
+
+/// Default value for the float comparison inside `EvolutionPoint`.
+const EP_CMP_ULPS: i64 = 64;
+
+/// A reference point in the evolution atlas.
+pub struct EvolutionPoint {
+    /// Evolution scale.
+    pub scale: f64,
+    /// Number of flavors
+    pub nf: i64,
+}
+
+/// 4D evolution operator.
+pub struct Operator {
+    pub op: Option<Array4<f64>>,
+}
+
+impl Operator {
+    /// Empty initializer.
+    pub fn new() -> Self {
+        Self { op: None }
+    }
+}
+
+impl Default for Operator {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl TryFrom<&Yaml> for EvolutionPoint {
+    type Error = EKOError;
+
+    /// Load from yaml.
+    fn try_from(yml: &Yaml) -> Result<Self> {
+        // work around float representation
+        let scale = yml["scale"].as_f64();
+        let scale = if scale.is_some() {
+            scale.ok_or(EKOError::KeyError(
+                "because failed to read scale as float".to_owned(),
+            ))?
+        } else {
+            yml["scale"].as_i64().ok_or(EKOError::KeyError(
+                "because failed to read scale as float from int".to_owned(),
+            ))? as f64
+        };
+        let nf = yml["nf"]
+            .as_i64()
+            .ok_or(EKOError::KeyError("because failed to read nf".to_owned()))?;
+        Ok(Self { scale, nf })
+    }
+}
+
+impl PartialEq for EvolutionPoint {
+    /// (Protected) comparator.
+    fn eq(&self, other: &Self) -> bool {
+        self.nf == other.nf && approx_eq!(f64, self.scale, other.scale, ulps = EP_CMP_ULPS)
+    }
+}
+
+impl Eq for EvolutionPoint {}
+
+/// EKO output
+pub struct EKO {
+    /// Working directory
+    path: PathBuf,
+    /// final operators
+    operators: inventory::Inventory<EvolutionPoint>,
+}
+
+/// Operators directory.
+const DIR_OPERATORS: &str = "operators/";
+
+impl EKO {
+    /// Check our working directory is safe.
+    fn check(&self) -> Result<()> {
+        let path_exists = self.path.try_exists().is_ok_and(|x| x);
+        if !path_exists {
+            return Err(EKOError::NoWorkingDir);
+        }
+        Ok(())
+    }
+
+    /// Remove the working directory.
+    pub fn destroy(&self) -> Result<()> {
+        self.check()?;
+        Ok(remove_dir_all(&self.path)?)
+    }
+
+    /// Write content back to an archive and destroy working directory.
+    pub fn write_and_destroy(&self, dst: PathBuf) -> Result<()> {
+        self.write(dst)?;
+        self.destroy()?;
+        Ok(())
+    }
+
+    /// Write content back to an archive.
+    pub fn write(&self, dst: PathBuf) -> Result<()> {
+        self.check()?;
+        // create writer
+        let dst_file = File::create(&dst)?;
+        let dst_file = BufWriter::with_capacity(128 * 1024, dst_file);
+        let mut ar = tar::Builder::new(dst_file);
+        // do it!
+        ar.append_dir_all(".", &self.path)?;
+        Ok(())
```

Comment:
```suggestion
        ar.append_dir_all(".", &self.path).into()
```

Here and in `.destroy()` you're actually receiving a different `Result` from yours. But the `thiserror` declaration above should implement somehow `From` for your `Result`, and I'm pretty sure there is no magic in `?` (other than early returning, and automated `.into()` invocation). Though I didn't tested (but the compiler will tell you immediately).

Reply:
`.into()` doesn't work either (same as ["plain" return](https://github.com/NNPDF/eko/pull/400#discussion_r1724616312) doesn't) - I changed to the shortest form `Ok(blub?)` here 8eb8775c